### DISCUSSION TITLE
[cc65] Sequence points?

### DISCUSSION THREAD

**Discussion by acqn** [cc65/cc65#1835] [General]
Given:
```c
exprA + (exprB, exprC);
```
C89 text: .

> **2.1.2.3 Program execution**
...
Evaluation of an expression may produce side effects. At certain specified points in the execution sequence called sequence points, _**all side effects of previous evaluations** shall be complete and no side effects of subsequent evaluations shall have taken place_. 

vs. C17 text:

> **5.1.2.3 Program execution**
3 ...
The presence of a sequence point between the evaluation of expressions A and B implies that _**every value computation and side effect associated with A** is sequenced before every value computation and side effect associated with B_.

So can we conclude that, in C89:
- both `exprA` and `exprB` are "sequenced before" `exprC`.
- `exprA` and `exprB` are unsequenced.

whilst in C17:
- only `exprB` is sequenced before `exprC`.
- `exprA` and `(exprB, exprC)` are unsequenced.

? (cc65 is currently implementing the "C89 way")

**Comment by mrdudz** [comment]
I have to admit i can't comment on this, the comma operator is something i totally ignore :) Is there another case where this matters?

> cc65 is currently implementing the "C89 way"

I'd say that pretty much already is the answer - or rather raises another question: "do we really want to deal with both implementations?" ie if STD=C89 then it must be this way, the C17 way could only be used for STD=CC65 right now.

**Comment by polluks** [comment]
@acqn 
Do you think C17 allows better code?

**Comment by mrdudz** [comment]
> @acqn Do you think C17 allows better code?

Oh please. This is a tiny detail that determines which order certain things happen.

**Comment by pmor13** [comment]
@acqn How side effects are relevant here? The semantics of comma operator (both C11 and C89) has no side effects.

**Comment by acqn** [comment]
> @acqn How side effects are relevant here? The semantics of comma operator (both C11 and C89) has no side effects.

There is a sequence point with the comma operator. It seems to me that:
```c
#include <stdio.h>

int foo(char x)
{
    printf("%c", x);
    return 0;
}

int main(void)
{
    foo('A') + (foo('B'), foo('C'));
    return 0;
}
```
is allowed to output `BCA` in C17 but never in C89.

**Comment by pmor13** [comment]
Re:
> So can we conclude that, in C89:
> both exprA and exprB are "sequenced before" exprC

[C89, 3.3 EXPRESSIONS](http://port70.net/~nsz/c/c89/c89-draft.html#2.1.2.3):
> Except as indicated by the syntax[27](http://port70.net/~nsz/c/c89/c89-draft.html#27) or otherwise specified later (for the function-call operator () , && , || , ?: , and comma operators), **the order of evaluation of subexpressions** and the order in which side effects take place are both **unspecified**.

This may be useful: https://en.cppreference.com/w/c/language/eval_order.

**Comment by acqn** [comment]
I see. So in C89 `BCA` is still allowed since the evaluation of `exprA` is not necessarily as "**previous**" as the standard requires (because its order is **unspecified** first of all).
> all side effects of _**previous** evaluations_

**Comment by mrdudz** [comment]
so this is a non issue? :)

**Comment by mrdudz** [comment]
moving to discussions
